### PR TITLE
feat: refreshValues api

### PR DESCRIPTION
**Pull request by blhoward2**
I'm attempting to expose the refreshValues() function to mqtt to permit a manual refresh of the dynamic values. To fix: #245 

I've already submitted a PR upstream making the function public. https://github.com/zwave-js/node-zwave-js/pull/1403

### CODE REVIEW

**Comment by blhoward2** [comment]
@AlCalzone @robertsLando I'd appreciate some help here. I can't seem to make this work. I've made the function in Node.js public, and I've tried every possible way that I can think of to access that function but none work. this.nodes.refreshValues(), this.controller.node.refreshValues(), this.controller.nodes.refreshValues(), etc. What am I doing wrong? 

I get success:false when I call it in MQTT, with a message that function is either undefined or that the property cannot be read from undefined.

**Comment by robertsLando** [comment]
@blhoward2 because `this.nodes` are not the driver nodes. In order to call that function you have to use `this.driver.controller.nodes`, also isn't that value node specific? If so even better to use `this.getNode(nodeId).refreshValues()`

**Comment by AlCalzone** [comment]
We should make it clear that calling this often is not a good idea because it might cause lots of traffic.

Additionally, we should make it possible to refresh single values, so applications can work around device quirks, e.g. not updating some values correctly.

**Comment by blhoward2** [comment]
> We should make it clear that calling this often is not a good idea because it might cause lots of traffic.
> 
> Additionally, we should make it possible to refresh single values, so applications can work around device quirks, e.g. not updating some values correctly.

I posted on the node pr too. I think having both is a good idea. I’m trying to duplicate ozw’s refreshDynamic service that updated all dynamic values, which I think this does. I agree it would be good to be able to get a single value too.

**Comment by blhoward2** [comment]
> @blhoward2 because `this.nodes` are not the driver nodes. In order to call that function you have to use `this.driver.controller.nodes`, also isn't that value node specific? If so even better to use `this.getNode(nodeId).refreshValues()`

Ok, I tried this in the new commit but I get "Cannot read property 'refreshValues' of undefined" in mqtt. Thoughts?

**Comment by robertsLando** [comment]
Are you passing the nodeid in mqtt payload args?

**Comment by blhoward2** [comment]
> Are you passing the nodeid in mqtt payload args?

Hmm, I am not. I can do so, I suppose.

File: lib/ZwaveClient.js [hunk 1]
```diff
@@ -1838,8 +1838,8 @@ ZwaveClient.prototype.getInfo = function () {
   return info
 }
 
-ZwaveClient.prototype.refreshValues = function () {
-  this.nodes.refreshValues()
+ZwaveClient.prototype.refreshValues = function (nodeId) {
+  this.getNode(nodeId).refreshValues()
```
Comment (AlCalzone) [inline lib/ZwaveClient.js:1842]: That method returns a Promise which could (theoretically) be rejected - should probably add a .catch()

**Comment by blhoward2** [comment]
> Are you passing the nodeid in mqtt payload args?

Same with the following:
```

{
  'nodeId': "11"
}
```

I tried it as both a string and an integer.

File: lib/ZwaveClient.js (hunk 1 above)
Comment (blhoward2) [reply lib/ZwaveClient.js]: Done in new commit

File: lib/ZwaveClient.js [hunk 2]
```diff
@@ -1838,6 +1838,12 @@ ZwaveClient.prototype.getInfo = function () {
   return info
 }
 
+ZwaveClient.prototype.refreshValues = function (nodeId) {
+  this.getNode(nodeId).refreshValues()
```
Comment (robertsLando) [inline lib/ZwaveClient.js]: Anyway I prefer something like: 

```js
ZwaveClient.prototype.refreshValues = async function (nodeId) {
  if (this.driver && !this.closed) {
    const zwaveNode = this.getNode(nodeId)

    return zwaveNode.refreshValues()
  }

  throw Error('Driver is closed')
}
```

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: lib/ZwaveClient.js (hunk 2 above)
Comment (blhoward2) [reply lib/ZwaveClient.js]: Should it be a throw or a catch per @AlCalzone comment?

**Comment by robertsLando** [comment]
@blhoward2 

```
{
  "args": [nodeId]
}
```

Check https://zwave-js.github.io/zwavejs2mqtt/#/guide/mqtt?id=zwave-apis

File: lib/ZwaveClient.js (hunk 2 above)
Comment (blhoward2) [reply lib/ZwaveClient.js]: also should it still return true?

**Comment by blhoward2** [comment]
> @blhoward2
> 
> ```
> {
>   "args": [nodeId]
> }
> ```
> 
> Check https://zwave-js.github.io/zwavejs2mqtt/#/guide/mqtt?id=zwave-apis

Is there noway to extra the nodeId from the mqtt command string? Like you do for getAssociations?

File: lib/ZwaveClient.js (hunk 2 above)
Comment (robertsLando) [reply lib/ZwaveClient.js]: This is how I have implemented all other methods if you check. It's correct to throw there

**Comment by AlCalzone** [comment]
Upstream PR merged

**Comment by robertsLando** [comment]
> Is there no way to capture the nodeId from the mqtt command string? Like you do for getAssociations? that would be easier to use since there are no other args

I don't understand what you mean here. What's the mqtt command string? All zwave apis and also others are called using the same way, I mean I simply apply the args coming from the mqtt to the function specified (if exists), making what you asked means create a big switch case where foreach api I fetch the required props from payload, this is just quicker

**Comment by blhoward2** [comment]
> > Is there no way to capture the nodeId from the mqtt command string? Like you do for getAssociations? that would be easier to use since there are no other args
> 
> I don't understand what you mean here. What's the mqtt command string? All zwave apis and also others are called using the same way, I mean I simply apply the args coming from the mqtt to the function specified (if exists), making what you asked means create a big switch case where foreach api I fetch the required props from payload, this is just quicker

Maybe I've got this all wrong. If I publish homeassistant/_CLIENTS/ZWAVE_GATEWAY-11/api/getAssociations/set with no payload it returns Success:true. Is that not right? Where 11 is the node name?

**Comment by robertsLando** [comment]
@blhoward2 YOu can even add it to UI, just add an entry here: 

https://github.com/blhoward2/zwavejs2mqtt/blob/add_refresh_values/src/components/ControlPanel.vue#L691

**Comment by blhoward2** [comment]
I think it is working now with the committed changes. I'll verify in a bit once the network is full up. I get a success message back in mqtt, though.

**Comment by blhoward2** [comment]
@zwave-js-bot fix lint

**Comment by blhoward2** [comment]
@zwave-js-bot fix lint

**Comment by blhoward2** [comment]
How come the bot isn't responding?

**Comment by blhoward2** [comment]
Confirmed working on multiple devices.

File: docs/guide/mqtt.md [hunk 3]
```diff
@@ -74,6 +74,7 @@ There are some custom apis that can be called that are not part of Zwave Client:
   - `lastUpdate`: Timestamp of latest event received from OZW
   - `status`: Client status. Could be: 'driverReady', 'connected', 'scanDone', 'driverFailed', 'closed'
   - `cntStatus`: Controller status received from ozw notifications controller command. If inclusion/exclusion is running it would be `Waiting`
+- `refreshValues`: Queries all actuator and sensor values of all node endpoints
```
Comment (robertsLando) [inline docs/guide/mqtt.md:77]: This is not needed as it is a zwave api

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: docs/guide/mqtt.md (hunk 3 above)
Comment (blhoward2) [reply docs/guide/mqtt.md:77]: How do people know it exists, then? Someone on slack yesterday asked if it existed as they'd been looking through the docs.

Comment (robertsLando) [reply docs/guide/mqtt.md:77]: because that api supports all zwave apis by default, in the list I only specify the ones that are not from zwave

Comment (robertsLando) [reply docs/guide/mqtt.md:77]: We make a list of the supported ones too, If you want to

Comment (blhoward2) [reply docs/guide/mqtt.md]: Might be a good idea. I think the mqtt documentation in general is a little hard to understand. May be worth having a second look at the whole thing with an eye towards an average user. I removed it from the PR for now.

**Comment by blhoward2** [comment]
Whoops, that deletes the whole thing. How do I undo a commit?

File: docs/guide/mqtt.md (hunk 3 above)
Comment (robertsLando) [reply docs/guide/mqtt.md]: @blhoward2 I'm open to suggestion/help about how to improve it :)

**Comment by robertsLando** [comment]
@blhoward2 `git revert 1e1ff05`

File: docs/guide/mqtt.md (hunk 3 above)
Comment (blhoward2) [reply docs/guide/mqtt.md]: mqtt is an odd beast. It is fairly advanced but every now and then a normal user needs to publish a message (and HA lets you do so within an automation as an action)

**Review by robertsLando — APPROVED** [review]
LGTM

**Comment by robertsLando** [comment]
@blhoward2 Need a new beta or other release before merging this. @AlCalzone

**Comment by blhoward2** [comment]
Understood

**Comment by AlCalzone** [comment]
6.0.0 is going out tonight if there is no critical bug

**Pull request merged**